### DISCUSSION TITLE
Data limitations or model tuning problems when predicting water quality (temperature, phosphorus, etc.)?

### DISCUSSION THREAD

**Discussion by Flash-Of-Thunder** [neuralhydrology/neuralhydrology#159] [Q&A]
Hello, I'm attempting to predict temperature (T), total phosphorus (TP), total nitrogen (TN), and total suspended solids (TSS) and am wondering if I have a data problem or a model tuning problem, any direct thoughts or resources would be highly appreciated.

**Context:**
I'm working within a specific subset of Gages-II basins that have high agricultural area (901 USGS basins). I've been able to train temperature fairly easily with the attached config file and got a NSE distribution (out-of-sample using k-fold=12) that was very respectable:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/94713715/9e069b10-9b89-4073-af6a-ff81813b7eff)

Here is an example of the simulations and observed data:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/94713715/33676964-55c5-4b4d-805b-7d531ca8ac01)

**Challenge**

When I try to do the same for TP, TN, or TSS my models end up being...pretty bad:

Total Phosphorus (~144 basins trained together: K-fold out-of-sample prediction)
![image](https://github.com/neuralhydrology/neuralhydrology/assets/94713715/d31c1d14-9fb5-4b79-8935-8b6dd028898d)

Total Nitrogen (Single LSTM: Fully In-sample)
![image](https://github.com/neuralhydrology/neuralhydrology/assets/94713715/e9e0cb30-75f5-4472-9717-d9a8cb9f8312)

My approach to parameter tuning is pretty manual, I've leveraged inputs of previous published research but have not tried a grid hyperparameter tunning scheme. 

I'm wondering if I'm off on some hyperparameters or if my model is actually missing key inputs. From the literature it some papers have more robust inputs (e.g. lagged TP/TN/TSS) but it also seems that some get good fit/prediction with little data and just meteorological variables: e.g. https://www.sciencedirect.com/science/article/pii/S0169772223001328?casa_token=ULoehRdliS8AAAAA:f61E1MxOU4al1DXuv4SV0ImK2EoCDc6Fdf84aFRjCL5i1xOs7Ikn7KI45gLJp7FAkV6dOp4Bpmo#s0010

From this paper above, they achieve great out-of-sample in time results for different water quality parameters. They train on ~3 years of continuous data e.g. ~1100 data points. By comparison my water quality data is more sparce, with the # of data points rarely exceeding 300 per basin. They also use a slightly different structure, including this "Dense" layer that I'm less familiar with. 
https://ars.els-cdn.com/content/image/1-s2.0-S0169772223001328-gr6_lrg.jpg

![image](https://github.com/neuralhydrology/neuralhydrology/assets/94713715/fb67e3fd-ea41-47a1-970a-5ed81fa451e7)

Attached are two examples of my config/yml files, for temperature I train ~200 basins simultaneously. For nutrients (TP/TN) I also tried doing single fully in-sample predictions, which is attached.  

[Temperature Config.txt](https://github.com/neuralhydrology/neuralhydrology/files/14143400/Temperature.Config.txt)
[Total Nitrogen Config.txt](https://github.com/neuralhydrology/neuralhydrology/files/14143401/Total.Nitrogen.Config.txt)

Another thing I'll note is that my models seem to really fail in predicting extremes, even my temperature model falls a little flat in this regard...

**Comment by kratzert** [comment]
Not really sure for what kind of advice you are looking for. I have never worked with this kind of data and this is all here just guessing into the blue. The only thing I noticed by glimpsing into your configs is that you train the total nitrogen model without static attributes. I am totally not surprised that this doesn't work. How would the model figure out why the different input samples (aka basins) behave differently, if you don't provide any static features?

**Reply by Flash-Of-Thunder** [reply]
Hello Frederik, yes good question. I'm looking for any kind of thoughts that'd direct my next dive into this analysis. Specifically, I'm trying to decide if I should focus on hyperparameter tuning to improve the model OR if the parameters are largely fine and the data is simply too scarce to fit anything accurately. 

You're response is helpful in getting me thinking about this dilemma. As for your specific point about no static attributes in the total nitrogen, that is because the total nitrogen is training on just 1 single basin. Apologies I could have cleaned up my config file more, as it may be misleading, I left:
```
# specify how many random basins to use for validation
validate_n_random_basins: 901
```

But it is only training on 1 basin. Previous runs trained on multiple basins did include static attributes, but I didn't change that line as I thought it doesn't matter. I.e. if I only have 1 basins and my config file says 901 to validate, it simply validates the max (1 basin). Let me know if this is incorrect. Again, it's a good point to spend more time cleaning up my config file before posting. 

If that's the only thing that pops out at you, then I'm inclined to think the config files are "okay" and that my data is trash. Perhaps I'll try running a similar analysis on the CAMELS dataset, it should be easier to fit in theory.
